### PR TITLE
core/mutex: docs: clarify usage of mutex_unlock().

### LINKED ISSUE #9594: Q: Is it OK to unlock a mutex on a different thread from which it was locked?
This question was triggered by the use of a mutex as a signalling device in #9588.

In many OSes (POSIX, Windows), mutexes cannot be released by a thread that did not lock them. RIOT seems to allow that, and thus RIOT's mutex seems to be a more general synchronization mechanism.

If this usage is supported, it should be documented and guaranteed to work. If it is not, there should be a warning to avoid that kind of usage.

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

Mutexes in RIOT can be locked by a thread an unlocked by other. This is not a very common feature and, given the lack of other synchronization primitives in `core`, a very useful one.

This change documents the behaviour. Additionally, now that it is written we are bound to respect this API.

### Issues/PRs references

Fixes #9594.

### CODE REVIEW

File: core/include/mutex.h [hunk 1]
```diff
@@ -106,6 +106,10 @@ static inline int mutex_trylock(mutex_t *mutex)
 /**
  * @brief Locks a mutex, blocking.
  *
+ * Mutexes are served in FIFO order. This ensures that, as long as no thread
```
Comment (kaspar030) [inline core/include/mutex.h]: I think is is more confusing than correct.
Actually, mutex waiters are served sorted by priority (highest priority / lowest priority value first), then FIFO.

File: core/include/mutex.h [hunk 2]
```diff
@@ -116,6 +120,13 @@ static inline void mutex_lock(mutex_t *mutex)
 /**
  * @brief Unlocks the mutex.
  *
+ * If there is more than one thread waiting for the mutex, the first one that
```
Comment (kaspar030) [inline core/include/mutex.h]: See above.

Comment (kaspar030) [reply core/include/mutex.h]: Correct would be: "Should there be other threads be waiting on the mutex, the mutex will stay locked and the highest-priority thread that has waited the longest will be woken up."

File: core/include/mutex.h [hunk 3]
```diff
@@ -116,6 +120,13 @@ static inline void mutex_lock(mutex_t *mutex)
...
+ * requested the mutex will be unlocked (i.e. first come, first serve).
+ *
+ * In contrast with other systems, a mutex in RIOT can be unlocked by a thread
```
Comment (kaspar030) [inline core/include/mutex.h]: How about just "A mutex can be unlocked by any thread or ISR."
There are "other systems" which behave simliar, e.g., Pythons Lock().

File: core/include/mutex.h (hunk 1 above)
Comment (jcarrano) [reply core/include/mutex.h]: Thanks for the clarification. I guess this confirms the behaviour is sufficiently complex as to need documentation.

Comment (kaspar030) [reply core/include/mutex.h]: Yup. Maybe even some module-level overview documentation?

**Comment by miri64** [comment]
Ping @jcarrano?

**Comment by jcarrano** [comment]
I reworded it, see if you like it.

File: core/include/mutex.h [hunk 4]
```diff
@@ -106,6 +106,12 @@ static inline int mutex_trylock(mutex_t *mutex)
 /**
  * @brief Locks a mutex, blocking.
  *
+ * Mutexes are served in FIFO order while respecting thread priority (i.e. a higher
+ * priority thread will be unlocked before a lower priority one even if it arrived
+ * later). As long as no thread holds the mutex indefinitely and as long as there
+ * are no deadlocks, all same-priority threads blocked on mutex_lock() should
+ * eventually unblock.
```
Comment (danpetry) [inline core/include/mutex.h]: I'm not sure about what the point being made in this sentence is? It sounds like you're just explaining what happens when a mutex is unlocked (i.e. waiting threads unblock)? Also, shouldn't _all_ threads blocked on the mutex unblock when it's unlocked?
(I've removed this sentence in the PR as AFAICT it doesn't add anything, please correct me if I'm wrong)

**Pull request closed without merging**